FreeRTOS/FreeRTOS
# Fix usGenerateChecksum on 8-bit platforms with odd-aligned buffers

Fix usGenerateChecksum on 8-bit platforms with odd-aligned buffers

Description
-----------
 + Fix casting of pointers to integers when pointers aren't 32-bits wide.
 + Fix generation of checksum on odd-aligned buffers with non-zero initialisation vector.

Test Steps
-----------
On avr8 the stack complains about invalid checksums for arriving UDP and TCP packets until this patch is applied.

Surprisingly, the stack can receive the DHCP packets necessary to obtain an IP address but all other UDP and TCP packets seem to fail the UDP or TCP checksum. The IP header checksum is not affected.

Reproduce:
TCP:
...by trying to telnet to a port on the target
`$ telnet 172.25.11.35 51`

UDP:
...by using netcat to send packets to the target
`echo "Hello" | nc -u 172.25.11.35 51`


I was running this on an ATMega1284 with an Microchip ENC28J60 combined MAC and PHY.


I reproduced the issue by compiling the usGenerateChecksum function on Linux-buster-x86_64 and passing it captured packets that I had pre-padded to be aligned on different boundaries. I can supply this code if desired.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
